docs/ds/dsu.md

```diff
@@ -1,150 +1,156 @@
 
 ```cpp
-int fa[MAXN];  // 记录某个人的爸爸是谁，特别规定，祖先的爸爸是他自己
+int fa[MAXN];  // record who is who's father, with special rules that the father of an ancestor is himself
 int find(int x) {
-  // 寻找x的祖先
-  if (fa[x] == x)  // 如果x是祖先则返回
+  // find the ancestors of x
+  if (fa[x] == x)  // return if x is an ancestor
     return x;
   else
-    return find(fa[x]);  // 如果不是则x的爸爸问x的爷爷
+    return find(fa[x]);  // if not, then x’s dad asks x’s grandpa
 }
 ```
 
-显然这样最终会返回 $x$ 的祖先。
+Obviously this will eventually return the ancestor of $x$ .
 
-### 路径压缩
+### Path compression
 
 这样的确可以达成目的，但是显然效率实在太低。为什么呢？因为我们使用了太多没用的信息，我的祖先是谁与我父亲是谁没什么关系，这样一层一层找太浪费时间，不如我直接当祖先的儿子，问一次就可以出结果了。甚至祖先是谁都无所谓，只要这个人可以代表我们家族就能得到想要的效果。 **把在路径上的每个节点都直接连接到根上** ，这就是路径压缩。
 
 此处给出一种 C++ 的参考实现：
 
+The above way can indeed achieve the goal, but obviously the efficiency is too low. Why? Because we use too much useless information. Who my ancestors are and who my father is has nothing to do with it, and it is a waste of time to find generaion by generation. It’s better to let "me" be the "son" of my ancestors and obtain results. It doesn't even matter who the ancestor is, as long as this person can represent the family. **Connecting each node on the path directly to the root**, and this is path compression.
+
+Here we offer a C++ implementation for your reference:
+
 ```cpp
 int find(int x) {
-  if (x != fa[x])  // x不是自身的父亲，即x不是该集合的代表
-    fa[x] = find(fa[x]);  // 查找x的祖先直到找到代表,于是顺手路径压缩
+  if (x != fa[x])  // x is not its own father, that is, x is not a representative of the set
+    fa[x] = find(fa[x]);  // find the ancestors of x until you find a representative, so the path is compressed
   return fa[x];
 }
 ```
 
-上两张图：
+See these two figures:
 
 ![p1](./images/dsu1.png)
 
-![p2](./images/dsu2.png)
+![p2](./images/dsu2-en.png)
 
-## 合并
+## Union
 
-宴会上，一个家族的祖先突然对另一个家族说：我们两个家族交情这么好，不如合成一家好了。另一个家族也欣然接受了。  
-我们之前说过，并不在意祖先究竟是谁，所以只要其中一个祖先变成另一个祖先的儿子就可以了。
+At the banquet, the ancestor of one family suddenly said to another family: Our two families have such a good relationship, so what about forming one family? The other family also gladly accepted.
+As we said before, we don’t care who the ancestors are, so as long as one of the ancestors becomes the son of the other ancestor.
 
-此处给出一种 C++ 的参考实现：
+Here we offer a C++ implementation for your reference:
 
 ```cpp
 void unionSet(int x, int y) {
-  // x 与 y 所在家族合并
+  // merge the families of x and y
   x = find(x);
   y = find(y);
-  if (x == y)  // 原本就在一个家族里就不管了
+  if (x == y)  // do not care if it was in the same family
     return;
-  fa[x] = y;  // 把 x 的祖先变成 y 的祖先的儿子
+  fa[x] = y;  // turn x's ancestor into y's ancestor's son
 }
 ```
 
-### 启发式合并（按秩合并）
+### Heuristic merge (merging by rank)
 
-一个祖先突然抖了个机灵：「你们家族人比较少，搬家到我们家族里比较方便，我们要是搬过去的话太费事了。」
+One of the ancestors suddenly had an interesting idea:"Your family has fewer people. It is more convenient to move to our family. It would be too much trouble if we move there."
```

Comment:
```suggestion
One of the ancestors suddenly had an interesting idea: "Your family has fewer people. It is more convenient to move to our family because it would be too troublesome if we move there."
```